File: ios/Livtet/Models/WizardData.swift
import Foundation
import LivtetKitFFI

/// Source of the cover image entered in Step 1 — Title and Cover.
///
/// `remote` is the only case the Phase 1 wizard can land in: the user has
/// either typed/pasted a URL, picked a search result, or selected a
/// photo from the OS picker. The `pendingLocal` and `downloaded` cases
/// are wired through the view-model state machine but their persistence
/// paths (the FFI download + `setEditionCover`) are intentionally not
/// exercised by Phase 1 — saving is disabled until the core/ FFI work
/// described in the Phase 2 spec lands.
enum CoverSource: Equatable {
    case remote(URL)
    case pendingLocal(uri: String, mimeType: String, byteSize: Int)
    case downloaded(localPath: String)

    var displayURL: URL? {
        switch self {
        case .remote(let url): return url
        case .pendingLocal(let uri, _, _): return URL(string: uri)
        case .downloaded(let localPath): return URL(fileURLWithPath: localPath)
        }
    }
}

struct WizardData: Equatable {
    var title: String = ""
    var description: String = ""
    var isbn: String = ""
    var publishedDate: String = ""
    var languageId: DbId? = nil
    var formatId: DbId? = nil
    var publisher: String = ""
    var authors: [AuthorEntry] = []
    var tags: [String] = []
    var genres: [String] = []
    var subjects: [String] = []
    var cover: CoverSource? = nil
    var searchQuery: String = ""
    var localDedupResults: [WorkSummary] = []
    var searchResults: [ProviderResult] = []

    /// Backward-compat accessor for the original `coverUrl: URL?` field.
    /// Older code paths (the Hub's "Cover Image" row, `CoverDetailView`'s
    /// prefilled text field) still read `data.coverUrl`; this lets them
    /// keep working without a separate stored field. New code should
    /// read `cover` directly and pattern-match on the `CoverSource` case.
    var coverUrl: URL? {
        get { cover?.displayURL }
        set {
            if let url = newValue {
                cover = .remote(url)
            } else {
                cover = nil
            }
        }
    }
}

enum WizardPage: Hashable {
    /// Linear 5-step flow pages — driven by the wizard router and the
    /// step views' `goToNext` / `goToBack` transitions.
    case titleAndCover
    case contributors
    case genres
    case subjects
    case tags
    case hub

    /// Hub field pages — never assigned to `currentPage` in the new
    /// flow, but kept in the enum so the existing `HubView` detail
    /// rows can switch on them when the user opens the "More options"
    /// hub from the Tags step. The legacy `cover` case is preserved
    /// for parity; the new Step 1 already gathers the cover before the
    /// user can reach the hub.
    case description, isbn, publishedDate, language, format, publisher, cover
}

struct AuthorEntry: Equatable, Identifiable {
    let id = UUID()
    var name: String = ""
    var role: String = "author"
}

struct ProviderResult: Equatable, Identifiable {
    let id = UUID()
    let title: String
    let authors: [String]
    let isbn: String?
    let year: Int?
    let publisher: String?
    let coverUrl: URL?
    let source: String
}

File: android/app/src/main/java/net/olamaelcu/livtet/wizard/WizardData.kt
package net.olamaelcu.livtet.wizard

/**
 * Source of the cover image entered in Step 1 — Title and Cover.
 *
 * [Remote] is the only case the Phase 1 wizard can land in: the user has
 * either typed/pasted a URL, picked a search result, or selected a photo
 * from the OS picker. The [PendingLocal] and [Downloaded] cases are
 * wired through the view-model state machine but their persistence paths
 * (the FFI download + `setEditionCover`) are intentionally not exercised
 * by Phase 1 — saving is disabled until the core/ FFI work described in
 * the Phase 2 spec lands.
 */
sealed class CoverSource {
    data class Remote(val url: String) : CoverSource()
    data class PendingLocal(val uri: String, val mimeType: String, val byteSize: Int) : CoverSource()
    data class Downloaded(val localPath: String) : CoverSource()

    val displayUrl: String?
        get() = when (this) {
            is Remote -> url
            is PendingLocal -> uri
            is Downloaded -> localPath
        }
}

data class WizardData(
    val title: String = "",
    val cover: CoverSource? = null,
    val searchQuery: String = "",
    val onlineResults: List<ProviderResult> = emptyList(),
    val contributors: List<ContributorEntry> = emptyList(),
    val genres: List<String> = emptyList(),
    val subjects: List<String> = emptyList(),
    val tags: List<String> = emptyList(),
)

data class ContributorEntry(
    val name: String,
    val role: String = "author",
)

data class ProviderResult(
    val title: String,
    val authors: List<String> = emptyList(),
    val isbn: String? = null,
    val year: Int? = null,
    val publisher: String? = null,
    val source: String = "",
    val coverUrl: String? = null,
)

enum class WizardStep {
    TITLE_AND_COVER,
    CONTRIBUTORS,
    GENRES,
    SUBJECTS,
    TAGS,
}

/**
 * Role options for the contributors step. "author" is the only role
 * the gate (see [AddBookWizardViewModel.canContinueFromContributors])
 * accepts as a "valid" contributor — every other role is allowed but
 * does not satisfy the gate on its own.
 */
val CONTRIBUTOR_ROLES = listOf("author", "illustrator", "translator", "narrator")
